mono/metadata/appdomain.c

```diff
@@ -706,56 +706,112 @@ mono_domain_has_type_resolve (MonoDomain *domain)
 
 /**
  * mono_domain_try_type_resolve:
- * \param domain application domainwhere the name where the type is going to be resolved
+ * \param domain application domain in which to resolve the type
  * \param name the name of the type to resolve or NULL.
- * \param tb A \c System.Reflection.Emit.TypeBuilder, used if name is NULL.
+ * \param typebuilder A \c System.Reflection.Emit.TypeBuilder, used if name is NULL.
  *
  * This routine invokes the internal \c System.AppDomain.DoTypeResolve and returns
- * the assembly that matches name.
- *
- * If \p name is null, the value of \c ((TypeBuilder)tb).FullName is used instead
+ * the assembly that matches name, or ((TypeBuilder)typebuilder).FullName.
  *
  * \returns A \c MonoReflectionAssembly or NULL if not found
  */
 MonoReflectionAssembly *
-mono_domain_try_type_resolve (MonoDomain *domain, char *name, MonoObject *tb)
+mono_domain_try_type_resolve (MonoDomain *domain, char *name, MonoObject *typebuilder)
 {
+	g_assert (domain);
+	g_assert (name || typebuilder);
+
 	MonoError error;
-	MonoReflectionAssembly *ret = mono_domain_try_type_resolve_checked (domain, name, tb, &error);
-	mono_error_cleanup (&error);
+	error_init (&error);
+
+	MonoReflectionAssembly * const ret = name
+		? mono_domain_try_type_resolve_name (domain, name, &error)
+		: mono_domain_try_type_resolve_typebuilder (domain, (MonoReflectionTypeBuilder *)typebuilder, &error);
 
+	mono_error_cleanup (&error);
 	return ret;
 }
 
+/**
+ * mono_class_get_appdomain_do_type_resolve_method:
+ *
+ * This routine returns System.AppDomain.DoTypeResolve.
+ */
+static MonoMethod *
+mono_class_get_appdomain_do_type_resolve_method (MonoError *error)
+{
+	static MonoMethod *method; // cache
+
+	if (method)
+		return method;
+
+	// not cached yet, fill cache under caller's lock
+
+	method = mono_class_get_method_from_name_checked (mono_class_get_appdomain_class (), "DoTypeResolve", -1, 0, error);
+
+	if (!mono_error_ok (error))
```

Comment:
`is_ok` is faster and preferable when used inside the runtime

Reply:
I not aware of entire error-handling-runtime and will look -- I imagine this is "macros have no linkage, only functions do" so outsiders must use functions, insiders can use macros. I was copying what I thought I'd seen aruond, will look again